Update Best Practices: Revise and Clarify "Avoid creating empty projects" item (#33918)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/best-practices/best_practices_structuring_builds.adoc

When you use the link:{kotlinDslPath}/gradle/org.gradle.api.initialization/-settings/include.html[Settings.include()] method to include a project in your Grade settings file, you typically include projects by supplying the directory name like `include("featureA")`.
This usage assumes that `featureA` is located at the root of your build.

> You can also include projects located in deeper subdirectories by specifying the complete path to the project using the `:` charactor as a path separator.

Reviewer: ```suggestion
You can include projects located in nested subdirectories by specifying their full project path by using `:` as a separator between path segments.
```

For instance, if project `search` was located in a subdirectory named `features`, itself located in a subdirectory named `subs`, you could call `include(":subs:features:search")` to include it.

> Nesting projects under a sensible hierarchical directory structure is a common practice in larger Gradle builds including more than a dozen or so projects.

Reviewer: You can also change it to `in larger Gradle builds with a dozen or more projects.` instead of my suggestion:

```suggestion
Nesting projects in a sensible hierarchical directory structure is common practice in larger Gradle builds with many projects. 
```

> Such a layout allows you to organize large builds and improve comprehensibility, versus having all your projects live at the root of the build.

Reviewer: ```suggestion
This approach helps organize large builds and improves comprehensibility, compared to placing all projects directly under the build root.
```

> However, without further configuration, Gradle will create empty projects for every element of a hierarchical path.

Reviewer: ```suggestion
However, without further configuration, Gradle will create empty projects for each element in the hierarchical path, even if some of those directories do not contain actual builds.
```

In the example above, Gradle will create a project named `:subs`, a project named `:subs:features`, and a project named `:subs:features:search`.

> This probably isn't intended, as you likely only want to add a `search` project.

Reviewer: ```suggestion
This behavior is usually not intended, as you likely only want to include the `search` project itself.
```

> Unused projects - even if empty - can surprise maintainers, clutter reporting, and make your build harder to understand.

Reviewer: ```suggestion
Unused projects - even if empty - can surprise maintainers, clutter reports, and make your build harder to understand.
```

> If plugins are applied using `allprojects { ... }` or `subprojects { ... }`, they will apply to each of these empty projects, which can also have performance impacts.

Reviewer: ```suggestion
They also introduce unintended side effects.
If you use `allprojects { ... }` or `subprojects { ... }`, plugins and configuration blocks will apply to every project, including the empty ones. 
This can degrade build performance.
```

> It can also be awkward to invoke tasks on nested projects, as you will have to use the full path to the project, like `gradle :subs:features:search:build`, instead of the shorter `gradle :search:build`.

Reviewer: ```suggestion
Additionally, invoking tasks on deeply nested projects requires using the full project path, such as `gradle :subs:features:search:build`, instead of the shorter `gradle :search:build`.
```

> To avoid these negatives when using a hierarchical project structure, you can provide a non-hierarchical name when including the project and be sure to explicitly set the link:{kotlinDslPath}/gradle/org.gradle.api/-project/get-project-dir.html[Project.projectDir] property of any projects included in nested directories:

Reviewer: ```suggestion
To avoid these downsides when using a hierarchical project structure, you can provide a flat name when including the project and explicitly set the link:{kotlinDslPath}/gradle/org.gradle.api/-project/get-project-dir.html[Project.projectDir] property for any projects located in nested directories:
```

```
include(':my-web-module')
project(':my-web-module').projectDir = file("subs/web/my-web-module")
```
This will prevent Gradle from creating empty projects for each element of the path.

NOTE: Always use an identical _logical_ project **name** and _physical_ project **location** to avoid confusion.  Don't include a project named `:search` and locate it at `features/ui/default-search-toolbar`, as this will lead to confusion about the location of the project.  Instead, locate this project at `features/ui/search`.

> Avoid unnecessarily deep directory structures - for builds containing only a few projects, it is better to keep the structure flat by locating all projects at the root of the build and avoid the need to explicitly set `projectDir` at all.

Reviewer: ```suggestion
You should avoid unnecessarily deep directory structures.
For builds containing only a few projects, it’s usually better to keep the structure flat by placing all projects at the root of the build.
This eliminates the need to explicitly set `projectDir`.
```

> Within the context of a particular build, it should always be obvious from the pathless project name where that project is located.

Reviewer: ```suggestion
Within the context of a particular build, the _pathless_ project name should clearly indicate where the project is located.
```

> If you find yourself dealing with ambiguity about project locations, consider flattening your directory structure to simplify it and avoid this.

Reviewer: ```suggestion
If you find yourself facing ambiguity about project locations, consider simplifying the directory layout by flattening the structure.
```

NOTE: You should typically keep your _logical_ project paths and _physical_ project locations aligned to avoid confusion.  Don't include a project named `:subs:features:search` and locate it at `features/ui/search`, as this will lead to confusion about the location of the project.

However, without further configuration, Gradle will create empty projects for every element of the path.
In the example above, Gradle will create a project named `:subs`, a project named `:subs:features`, and a project named `:subs:features:search`.
This probably isn't intended, as you likely only want to include the `search` project.
Additional projects - even if empty - will slow down your build, clutter reporting, and make your build harder to understand.

To avoid this, be sure to explicitly set the link:{groovyDslPath}/org.gradle.api.Project.html#org.gradle.api.Project:projectDir[Project.projectDir] property of any projects included in nested directories: `project(':my-web-module').projectDir = file("subs/web/my-web-module"`).

> As an added benefit, this will allow you to reference the project by the shorter name `:my-web-module` instead of the full logical path `:subs:web:my-web-module`.

Reviewer: I thought we wanted to either get rid of this sentence altogether or add a disclaimer about how it's still better to use the full format `:subs:web:my-web-module` so you don't forget "where" the project is located.
